cmiyc/dataset_utils.py

```diff
@@ -27,6 +37,166 @@ def load_clean_train(sig_type='all', sig_id='all', id_as_label='false'):
     labels = 'sig_id' if id_as_label else 'label'
     return np.vstack(df['sig'].to_numpy()), df[labels].to_numpy().astype('int')
 
-def get_unique_sig_ids():
-    df = pd.read_pickle(pre.PATH_TRAIN)
-    return sorted(df['sig_id'].unique())
\ No newline at end of file
+def load_clean_train_test(vae_sig_type='genuine', sig_id=1, id_as_label='false', frac=0.5, random_state=4):
+    '''
+    For a single sig_id, splits out train and test data as follows:
+
+    train: frac * {genuine examples from the Train folder}
+    test: (1-frac) * {genuine examples from the Train folder} + 
+                      {fake examples from the Train folder}
+
+    Args:
+
+    vae_sig_type: the signature type that is used to train the VAE. 
+    (default = 'genuine', reverse the description above if not 'genuine')
+    sig_id: cannot be 'all' here. Must be int
+
+    Random state is set to 4 for reproducibility.
+
+    Returns:
+    x_train, y_train, x_test, y_test
+    
+    '''
+
+    sig_id = [sig_id] if isinstance(sig_id, int) else sig_id
+
+    df = pd.read_pickle(pre.PATH_ALL)
+
+    df = df[df['sig_id'].isin(sig_id)]
+
+    if vae_sig_type == 'genuine':
+        vae_df = df[df['label'] == 1]
+        exp_df = df[df['label'] == 0]
+    elif vae_sig_type == 'forgery':
+        vae_df = df[df['label'] == 0]
+        exp_df = df[df['label'] == 1]
+
+    # split into:
+    # - A train set of genuines for the VAE
+    # - A test set of genuines + all fakes for the Experiment test
+    vae_df_train, vae_df_genuine_test = train_test_split(vae_df, test_size=1.0-frac, random_state=random_state)
+    exp_df_test = exp_df.append(vae_df_genuine_test)
+    
+    # Shuffle Experiment test set
+    exp_df_test = exp_df_test.sample(frac=1).reset_index(drop=True)
+
+    # Unit test to make sure our splits are sound
+    split_test(vae_df_train, exp_df_test, sig_id, vae_sig_type, verbose=False)
+
+    if id_as_label:
+        labels = 'sig_id'
+    else: 
+        labels = 'label' 
+
+    x_train = np.vstack(vae_df_train['sig'].to_numpy())
+    y_train = vae_df_train[labels].to_numpy().astype('int')
+
+    assert list(set(y_train)) == [TYPE_MAP[vae_sig_type]], "y_train contains bad values: {}".format(list(set(y_train)))
+
+    x_test = np.vstack(exp_df_test['sig'].to_numpy())
+    y_test = exp_df_test[labels].to_numpy().astype('int')
+
+    assert (1 in list(set(y_test)) and 0 in list(set(y_test))), "y_test contains bad values: {}".format(list(set(y_test)))
```

Comment:
What does this assertion do? It seems like it would trigger when `y_test` contains only 0s or 1s?

Reply:
check that y_test contains both 1 (genuine) and 0 (fakes), otherwise throw AssertionError

Reply:
and yeah...it would be bad to have a test set that only has a certain class, right?

Reply:
Oh yeah you're right! We want to make sure the test set has both class :+1: